jcl/src/java.base/share/classes/java/lang/StringBuilder.java

```diff
@@ -2261,7 +2261,43 @@ public StringBuilder append(CharSequence sequence, int start, int end) {
 	if (sequence == null) {
 		return append(String.valueOf(sequence), start, end);
 	} else if (sequence instanceof String) {
-		return append(((String)sequence).substring(start, end));
+		if (start >= 0 && end >= 0 && start <= end && end <= ((String) sequence).lengthInternal()) {
+			int currentLength = lengthInternal();
+			int currentCapacity = capacityInternal();
+			int newLength = lengthInternal() + end - start;
+			int sbBase = currentLength - start;
+			String sequenceString = (String) sequence;
+
+			if (String.enableCompression) {
+				// Check if the StringBuilder is compressed
+				if (count >= 0) {
+					decompress(newLength);
+				}
+
+				if (newLength > currentCapacity) {
```

Comment:
`currentCapacity` may not be correct here if a `decompress()` occurred.  Because this uses `ensureCapacityImpl`, this may result in additional array copies